scalableminds/webknossos
# Refactor Task Creation in Backend

Some clean up for the task creation code.
The complexity of task creation lies in the following requirements
 - Tasks can be created either from base annotation, from uploaded files, or plain. Creates a lot of different codepaths
 - Errors should be reported per task. Creates the need to carry around errors from previous steps
 - Perf should be acceptable (Can’t just run everything per task, need to do some things batched, this simplification for the above requirement is not feasible)

That is why this cleanup did not achieve great breakthroughs. However, I
 - renamed some methods to make their meaning clearer
 - moved some functionality into specialized case classes
 - moved the private methods of TaskController to new TaskCreationService
 - extracted some parts of long methods to separate named methods
 - moved methdos defined within other methods out
 - reordered methods in the code to reflect calling structure
 - added some comments
 - added more type hints and applied intellij style hints for a few more files

I hope this will help future understanding and edits of this code.

### URL of deployed dev instance (used for testing):
- https://refactortaskcreation.webknossos.xyz

### Steps to test:
- create some tasks (bulk, from files, plain)
- those tasks should be created and requestable
- try to create some tasks with broken input (e.g. from broken nmls)
- error messages should be readable

### Issues:
- fixes #4727 

------
